File: server/java/codegen-spring/src/main/java/com/metaobjects/generator/spring/runtime/FilterParseResult.java
package com.metaobjects.generator.spring.runtime;

import java.util.List;
import java.util.Objects;

/**
 * Outcome of {@link FilterParser#parse}: either a list of validated
 * predicates ({@code error == null}), or a cross-port error envelope key
 * ({@code predicates} is empty + {@code error} is one of
 * {@code invalid_filter_field} / {@code invalid_filter_op} /
 * {@code invalid_filter_value}) plus the {@code field} it is about.
 *
 * <p>{@code field} is non-null exactly when {@code error} is — the cross-port
 * envelope names the rejected field so a caller sending several filters does
 * not have to guess which one failed. The generated controller reads it
 * straight into the 400 body, which is why {@link #err} rejects a null field
 * at construction rather than letting it surface as a 500 from {@code Map.of}.
 */
public record FilterParseResult(List<FilterPredicate> predicates, String error, String field) {

    /** Success path: a list of (possibly empty) predicates. */
    public static FilterParseResult ok(List<FilterPredicate> predicates) {
        return new FilterParseResult(predicates, null, null);
    }

    /** Error path: cross-port envelope key + the offending field (no predicates). */
    public static FilterParseResult err(String error, String field) {
        return new FilterParseResult(List.of(), error, Objects.requireNonNull(field, "field"));
    }
}

File: server/csharp/MetaObjects.Codegen/Runtime/FilterParseResult.cs
// FilterParseResult — outcome of FilterParser.Parse: either a list of
// validated predicates (success) or an error envelope key (one of
// invalid_filter_field / invalid_filter_op / invalid_filter_value).
//
// Part of the codegen runtime surface; mirrors Java codegen-spring's
// runtime/FilterParseResult.java.

namespace MetaObjects.Codegen.Runtime;

/// <summary>
/// Result of <see cref="FilterParser.Parse"/>. Exactly one of
/// <see cref="Predicates"/> (success path) or <see cref="ErrorEnvelope"/>
/// (failure path) is meaningful; on failure <see cref="Predicates"/> is empty.
/// </summary>
/// <param name="Predicates">The validated + coerced predicates on success; empty on failure.</param>
/// <param name="ErrorEnvelope">
/// Cross-port error envelope key (<c>invalid_filter_field</c> /
/// <c>invalid_filter_op</c> / <c>invalid_filter_value</c>) on failure;
/// <see langword="null"/> on success.
/// </param>
/// <param name="Field">
/// The offending filter field on failure; <see langword="null"/> on success.
/// REQUIRED on the wire for every contract envelope that is about a field
/// (<c>invalid_filter_field</c> / <c>invalid_filter_op</c> /
/// <c>invalid_filter_value</c>) — the caller must not have to guess which of
/// several filters was rejected.
/// </param>
public sealed record FilterParseResult(
    System.Collections.Generic.IReadOnlyList<FilterPredicate> Predicates,
    string? ErrorEnvelope,
    string? Field = null)
{
    /// <summary>Success result with the given predicates (may be empty).</summary>
    public static FilterParseResult Ok(System.Collections.Generic.IReadOnlyList<FilterPredicate> predicates) =>
        new(predicates, null);

    /// <summary>Failure result with the cross-port error envelope key + offending field.</summary>
    public static FilterParseResult Err(string envelope, string field) =>
        new(System.Array.Empty<FilterPredicate>(), envelope, field);
}
